File: benchmarks/fib/fib.rb
# Ruby benchmark: naive recursive Fibonacci
# Usage: ruby fib.rb [n] [iters], defaults n=35 iters=10

def fib(n)
  return n if n <= 1
  fib(n - 1) + fib(n - 2)
end

n = 35
if ARGV.length > 0
  n = ARGV[0].to_i
  n = 35 if n <= 0 || n > 45
end
iters = 10
if ARGV.length > 1
  iters = ARGV[1].to_i
  iters = 1 if iters < 1
end

# warm-up
fib(n)

result = 0
total_ns = 0

iters.times do
  t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
  result += fib(n)
  t1 = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
  total_ns += t1 - t0
end

mean_ns = total_ns.to_f / iters
puts "result: #{result}"
puts "[@bench] mean #{format('%.1f', mean_ns)} ns (#{iters} iterations)"

File: benchmarks/fib/fib.py
"""Python benchmark: naive recursive Fibonacci
Usage: python fib.py [n] [iters], defaults n=35 iters=10
"""
import time
import sys
sys.setrecursionlimit(1000000)


def fib(n: int) -> int:
    if n <= 1:
        return n
    return fib(n - 1) + fib(n - 2)


def main() -> None:
    n = 35
    if len(sys.argv) > 1:
        n = int(sys.argv[1])
        if n <= 0 or n > 45:
            n = 35
    iters = 10
    if len(sys.argv) > 2:
        iters = int(sys.argv[2])

    # warm-up
    fib(n)

    result = 0
    total_ns = 0

    for _ in range(iters):
        t0 = time.perf_counter_ns()
        result += fib(n)
        t1 = time.perf_counter_ns()
        total_ns += t1 - t0

    mean_ns = total_ns / iters
    print(f"result: {result}")
    print(f"[@bench] mean {mean_ns:.1f} ns ({iters} iterations)")


if __name__ == "__main__":
    main()
